storage/sql/template.go

```diff
@@ -145,34 +199,167 @@ func (rs *RecordStream) Close() error {
 	return nil
 }
 
-func (rs *RecordStream) Next(ctx context.Context) (*execution.Record, error) {
+func (rs *RecordStream) RunWorker(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			rs.workerCloseErrChan <- ctx.Err()
+		default:
+		}
+
+		tx := rs.stateStorage.BeginTransaction().WithPrefix(rs.streamID.AsPrefix())
+
+		err := rs.RunWorkerInternal(ctx, tx)
+		if errors.Cause(err) == execution.ErrNewTransactionRequired {
+			tx.Abort()
+			continue
+		} else if waitableError := execution.GetErrWaitForChanges(err); waitableError != nil {
+			tx.Abort()
+			err = waitableError.ListenForChanges(ctx)
+			if err != nil {
+				log.Println("sql worker: couldn't listen for changes: ", err)
+			}
+			err = waitableError.Close()
+			if err != nil {
+				log.Println("sql worker: couldn't close storage changes subscription: ", err)
+			}
+			continue
+		} else if err != nil {
+			tx.Abort()
+			log.Printf("sql worker: error running sql read batch worker: %s, reinitializing from storage", err)
+			tx := rs.stateStorage.BeginTransaction().WithPrefix(rs.streamID.AsPrefix())
+			if err := rs.loadOffset(tx); err != nil {
```

Comment:
You have to create a new rs.rows here based on this offset. Overall I think it would be the cleanest if you just create two for loops (like I wrote you on messenger), where the outer creates the rows object for the current offset, and the inner just spins calling RunWorkerInternal.